### PR TITLE
Feature/ker/basic zmq rpc

### PR DESCRIPTION
**Pull request by keriksson-rosenqvist-oqc**
Simple ZMQ implementation of RPC.

### CODE REVIEW

File: src/QAT_RPC/qat_rpc/zmq/receiver.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    hw = None
+    if (calibration_file := os.getenv("TOSHIKO_CAL")) is not None:
+        calibration_file = Path(calibration_file)
+        if not calibration_file.is_absolute() and not calibration_file.is_file():
+            calibration_file = Path(Path(__file__).parent, calibration_file)
+        if not calibration_file.is_file():
+            raise ValueError(f"No such file: {calibration_file}")
+        log.info(f"Loading: {calibration_file} ")
+        hw = Calibratable.load_calibration_from_file(str(calibration_file))
+        log.debug("Loaded")
```
Comment (jfriel-oqc) [inline src/QAT_RPC/qat_rpc/zmq/receiver.py:36]: loaded what?

**Review by jfriel-oqc — APPROVED** [review]
approving, let's get this merged and start working on it from there

**Comment by keriksson-rosenqvist-oqc** [comment]
I have read the Contributor License Agreement and I hereby accept the Terms.

File: src/QAT_RPC/qat_rpc/zmq/receiver.py (hunk 1 above)
Comment (hamidelmaazouz) [reply src/QAT_RPC/qat_rpc/zmq/receiver.py:36]: @keriksson-rosenqvist Consider a context here like we do for reporting time metrics in QAT

File: src/QAT_RPC/qat_rpc/zmq/qat_commands.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import argparse
+from pathlib import Path
+
+from qat_rpc.zmq.wrappers import ZMQClient
+
+
+parser = argparse.ArgumentParser(prog="QAT submission service", description="Submit your QASM or QIR program to QAT.")
+parser.add_argument("program", type=str, help="Program string or path to program file.")
```
Comment (hamidelmaazouz) [inline src/QAT_RPC/qat_rpc/zmq/qat_commands.py:8]: I presume `program` doesn't need a flag here right ? sth like `runner --config <config> <program>` ?

File: src/QAT_RPC/qat_rpc/zmq/wrappers.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        self._context = zmq.Context()
+        self._socket = self._context.socket(socket_type)
+        self._timeout = 30.0
+        self._protocol = "tcp"
+        self._ip_address = "127.0.0.1"
+        self._port = "5556"
+
+    @property
+    def address(self):
+        return f"{self._protocol}://{self._ip_address}:{self._port}"
```
Comment (hamidelmaazouz) [inline src/QAT_RPC/qat_rpc/zmq/wrappers.py:24]: minor: would've leaned on the URL/URI structures here

File: src/QAT_RPC/qat_rpc/zmq/wrappers.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+    def _check_recieved(self):
```
Comment (hamidelmaazouz) [inline src/QAT_RPC/qat_rpc/zmq/wrappers.py:26]: you probably also want a retry mechanism here similar to _send() ?

File: src/tests/zmq/test_wrappers.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    config0.results_format.binary_count()
+    config0.repeats = 100
+    thread00 = threading.Thread(
+        target=execute_and_check_result,
+        args=(client0, program, config0, {"c": { "00": 100}}),
+    )
+    thread01 = threading.Thread(
+        target=execute_and_check_result,
+        args=(client0, program, config0, {"c": { "00": 100}}),
+    )
```
Comment (hamidelmaazouz) [inline src/tests/zmq/test_wrappers.py:69]: any thoughts how useful would a thread pool be ? to easily scale the number of threads and stress test the clients ?

File: pyproject.toml [hunk 6]
```diff
@@ -19,6 +24,9 @@ optional = true
 [tool.poetry.group.licenses.dependencies]
 pip-licenses = "^3.5.3"
 
+[tool.poetry.scripts]
+qat_comexe="qat_rpc.zmq.qat_commands:qat_run"
```
Comment (hamidelmaazouz) [inline pyproject.toml:28]: not the most intuitive name ^^, would've called it `qat_run` ^^

**Review by hamidelmaazouz — COMMENTED** [review]
Sorry, my comments came on late, leaving them here for next iteration

**Pull request merged**